### ISSUE TITLE
It's possible for `dcel_operations::split_edge` to violate the ccw requirement.

### ISSUE THREAD

**Issue by dylanowen** [Stoeoef/spade#96]
This results in a panic when legalizing edges:
```
Point2 { x: 462.12845, y: 12.433466 } Point2 { x: 462.03986, y: 12.483034 } Point2 { x: 462.1012, y: 12.448717 } wasn't CCW
stack backtrace:
 0: rust_begin_unwind
           at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:597:5
 1: core::panicking::panic_fmt
           at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:72:14
 2: spade::delaunay_core::triangulation_ext::TriangulationExt::legalize_edge
           at ./src/delaunay_core/triangulation_ext.rs:400:21
 3: spade::delaunay_core::triangulation_ext::TriangulationExt::legalize_vertex
           at ./src/delaunay_core/triangulation_ext.rs:341:13
 4: spade::delaunay_core::refinement::<impl spade::cdt::ConstrainedDelaunayTriangulation<V,DE,UE,F,L>>::resolve_encroachment
           at ./src/delaunay_core/refinement.rs:853:9
 5: spade::delaunay_core::refinement::<impl spade::cdt::ConstrainedDelaunayTriangulation<V,DE,UE,F,L>>::refine
           at ./src/delaunay_core/refinement.rs:575:29
 6: spade::delaunay_core::refinement::test::failing_refinement
           at ./src/delaunay_core/refinement.rs:1081:22
 7: spade::delaunay_core::refinement::test::failing_refinement::{{closure}}
           at ./src/delaunay_core/refinement.rs:1062:32
```

You can recreate the bug by doing this refinement:
```rust
#[rustfmt::skip]
#[test]
fn failing_refinement() -> Result<(), InsertionError> {
    let mut cdt = ConstrainedDelaunayTriangulation::<Point2<f32>>::new();

    for [start, end] in [
        [Point2 { x: -50.023544, y: -25.29227 }, Point2 { x: 754.23883, y: -25.29227 }],
        [Point2 { x: 754.23883, y: 508.68994 }, Point2 { x: -50.023544, y: 508.68994 }],
        [Point2 { x: -44.20742, y: 316.5185 }, Point2 { x: -50.023544, y: 318.19534 }],
        [Point2 { x: 11.666269, y: 339.4947 }, Point2 { x: 15.110367, y: 335.44138 }],
        [Point2 { x: 335.06403, y: 122.91455 }, Point2 { x: 340.15436, y: 132.04283 }],
        [Point2 { x: 446.92468, y: -6.7025666 }, Point2 { x: 458.70944, y: 14.341333 }],
        [Point2 { x: 458.70944, y: 14.341333 }, Point2 { x: 471.58313, y: 7.1453195 }],
        [Point2 { x: 467.80966, y: 0.40460825 }, Point2 { x: 468.6454, y: -0.061800003 }],
        [Point2 { x: 464.55957, y: -7.3688636 }, Point2 { x: 465.48816, y: -7.890797 }],
        [Point2 { x: 465.48816, y: -7.890797 }, Point2 { x: 461.57117, y: -14.898027 }],
        [Point2 { x: 465.42877, y: 10.587858 }, Point2 { x: 453.93112, y: 17.01763 }],
    ] {
        cdt.add_constraint_edge(start, end)?;
    }

    let result = cdt.refine(RefinementParameters::default());

    Ok(())
}
```
    
We panic in `legalize_edge` when checking ccw. But adding this to the `split_edge` function panics much sooner:
```rust
debug_assert!(math::is_ordered_ccw(
    dcel.vertex(v0).position(),
    dcel.vertex(edge0.origin).position(),
    dcel.vertex(edge1.origin).position(),
));
debug_assert!(math::is_ordered_ccw(
    dcel.vertex(v0).position(),
    dcel.vertex(edge1.origin).position(),
    dcel.vertex(edge2.origin).position(),
));
debug_assert!(math::is_ordered_ccw(
    dcel.vertex(v0).position(),
    dcel.vertex(edge2.origin).position(),
    dcel.vertex(edge3.origin).position(),
));
debug_assert!(math::is_ordered_ccw(
    dcel.vertex(v0).position(),
    dcel.vertex(edge3.origin).position(),
    dcel.vertex(edge0.origin).position(),
));
```

The problem definitely seems to arise from float precision issues which makes this kind of tedious to fix. I definitely shouldn't be doing a triangulation on points this close to a line in my own code but for this library I'm not sure what would be the best way to solve this.

**Comment by Stoeoef** [comment]
That's unexpected, thanks for the Report. Spare _should_ be able to Deal with FP inaccurracies due to its usage of the robust dependency. I'll take a look at this late!

**Comment by dylanowen** [comment]
I was thinking, we could just double check the ccw property before inserting edges here. I wasn't sure if there was a better way though, or if this problem could crop up in the other similar functions.

On an unrelated note the documentation diagrams are very nice for understanding the code base.

**Comment by Stoeoef** [comment]
> On an unrelated note the documentation diagrams are very nice for understanding the code base.

Thanks! Glad to hear they don't just help me, but also others :)

**Comment by Stoeoef** [comment]
Fixed with release 2.4.1, now on crates.io

The "solution" will simply skip any vertex that would violate the CCW requirement. I think that should be fine in practice since, if an input geometry suffers from this effect, the result is probably not good in either case.

Thanks for the detailed report and clear instructions for reproducing the issues - this made it really easy to analyse the issue and improve on it!

**Comment by dylanowen** [comment]
Thanks for the quick fix!
